agents/locobot/locobot_agent.py

```diff
@@ -136,7 +146,33 @@ def init_perception(self):
         if not hasattr(self, "perception_modules"):
             self.perception_modules = {}
         self.perception_modules["self"] = SelfPerception(self)
-        self.perception_modules["vision"] = Perception(self, self.opts.perception_model_dir)
+        self.perception_modules["vision"] = Perception(self.opts.perception_model_dir)
+
+    def perceive(self, force=False):
+        self.perception_modules["self"].perceive(force=force)
+
+
+        rgb_depth = self.mover.get_rgb_depth()
+        xyz = self.mover.get_base_pos_in_canonical_coords()
+        x, y, yaw = xyz
+        sio.emit("map", {
+            "x": x,
+            "y": y,
+            "yaw": yaw,
+            "map": self.mover.get_obstacles_in_canonical_coords()
+        })
+
+        previous_objects = DetectedObjectNode.get_all(self.memory)
+        new_state = self.perception_modules["vision"].perceive(rgb_depth,
```

Comment:
replace this with 
```
new_objects, updated_objects = = self.perception_modules["vision"].perceive(...
```

That way you can also get rid of `if new_state is not None:` statement below.

Reply:
I can't do that I think, because we return `None` when there are no ready `detections`.

See for example:

```
In [1]: def foo(x):
   ...:     if x == 1:
   ...:         return None
   ...:     else:
   ...:         return 1, 2
   ...:

In [2]: a, b = foo(1)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-2f6254e10c77> in <module>
----> 1 a, b = foo(1)

TypeError: cannot unpack non-iterable NoneType object

In [3]: a, b = foo(2)
```

Reply:
Returning `None, None` would fix that? Don't have a strong opinion, just was thinking of cutting down on lines.